### PR TITLE
[3.1] Fix really bad bug in .rotate

### PR DESCRIPTION
**Pull request by nabobalis**
Yes the branch name is wrong.

### CODE REVIEW

File: setup.cfg
Comment (nabobalis) [inline setup.cfg:51]: Either this or we need to < numpy to below 1.23.0

File: tox.ini
Comment (nabobalis) [inline tox.ini:69]: 4.2.0 and recent numpys arent happy, so I decided to bump this instead.

File: setup.cfg
Comment (nabobalis) [inline setup.cfg:187]: I am ignoring these but I don't know if they are important.

File: setup.cfg
Comment (dstansby) [inline setup.cfg:27]: Did we already support `3.10` with `3.1.x` and just miss this?

File: setup.cfg
Comment (dstansby) [inline setup.cfg:186]: Can we try and add these to individual tests instead of globally? That way it makes it easy to see that they're expected warnings in context.

File: sunpy/data/_sample.py
Comment (dstansby) [inline sunpy/data/_sample.py:10]: Does sample data work without making these changes? If so I think we should either leave the sample data alone, or if making these changes put them in a single commit with a meaningful commit message.

File: sunpy/util/parfive_helpers.py
Comment (dstansby) [inline sunpy/util/parfive_helpers.py:11]: To avoid making lots of changes, I think we should just pin the major version of parfive to `<2` if it's not compatible with `3.1.x`.

**Review by dstansby — COMMENTED** [review]
Given we're doing this I think it's prudent to make the set of changes absoutely minimal to avoid us tripping up on anything else, hence a lot of my review comments. Since we want to be careful I also think it pays to have a clean and clear commit history - could you clean it up so there's one change/commit with meaninful messages?

File: setup.cfg
Comment (ConorMacBride) [reply setup.cfg:119]: I think this ini option was added by mistake because someone had an extra non-sunpy pytest plugin installed which gave a warning if this ini option was missing.

File: sunpy/data/_sample.py
Comment (ConorMacBride) [reply sunpy/data/_sample.py:10]: Without these sample data changes it should fall back to the data.sunpy.org endpoint when we rename the sample data repo.

Comment (Cadair) [reply sunpy/data/_sample.py:10]: Given we are doing this, I would rather add this URL just to save me a few $ on the AWS bill. +1 on the history tho.

File: setup.cfg
Comment (dstansby) [reply setup.cfg:52]: How so? Shouldn't we be able to run fine with the minimum version (0.16.0) here?

**Comment by nabobalis** [comment]
The remaining CI fails are I think ok. We will bypass those builds for the release.

File: setup.cfg
Comment (nabobalis) [reply setup.cfg]: Ill remove it again.

**Comment by ConorMacBride** [comment]
Do these image differences all look fine? https://output.circle-artifacts.com/output/job/e250d8fb-ebf1-49cf-a9d4-7a61c8093682/artifacts/0/.tmp/py38-figure/figure_test_images/fig_comparison.html?f=image-diff&c=and

I realise the baselines for for 4.0, but are different croppings of rotations expected?

Also, I assume the oldestdeps failure is fine? https://github.com/sunpy/sunpy/runs/7417596641?check_suite_focus=true (This test was modified in #5867)

**Comment by nabobalis** [comment]
The figure comparisions are with main I assume? 

I think that oldest deps failure is ok.

**Comment by ConorMacBride** [comment]
> The figure comparisions are with main I assume?

Yes, that's what I meant sorry

> I think that oldest deps failure is ok.

Great!

**Comment by nabobalis** [comment]
We will need several people to double check the figure tests figures are ok.

**Review by dstansby — APPROVED** [review]
Thanks for putting this together! I think the figure test differences are fine, and all the code changes look fine too. This should definnitely get a second approval before merging.

File: changelog/6340.bugfix.rst
Comment (ayshih) [inline changelog/6340.bugfix.rst:1]:
```suggestion
Fixed an erroneous transpose of the image data by :meth:`sunpy.map.GenericMap.rotate`.
```
Not only does "transposing twice" not really accurately describe the bug, it reads silly since transposing twice would be the same as a no-op.

File: sunpy/image/tests/test_transform.py
Comment (ayshih) [inline sunpy/image/tests/test_transform.py:69]:
```suggestion
```

**Review by ayshih — APPROVED** [review]
Minor fixes needed, otherwise looks good to me

File: changelog/6340.bugfix.rst
Comment (nabobalis) [reply changelog/6340.bugfix.rst:1]: Fair point!

**Comment by nabobalis** [comment]
Thanks everyone, when the CI passes again, I will merge.

**Comment by wtbarnes** [comment]
Since we're here, can we also pin parfive to `2.0`? I'm getting this exception with `2.0.x` when doing `Fido.fetch` on 3.1.7,

```python traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 file = Fido.fetch(q)

File ~/anaconda/envs/eimason-sunpy/lib/python3.9/site-packages/sunpy/net/fido_factory.py:397, in UnifiedDownloaderFactory.fetch(self, path, max_conn, progress, overwrite, downloader, *query_results, **kwargs)
    395         max_conn = 1
    396         kwargs['max_splits'] = 1
--> 397     downloader = Downloader(max_conn=max_conn, progress=progress, overwrite=overwrite)
    398 elif not isinstance(downloader, parfive.Downloader):
    399     raise TypeError("The downloader argument must be a parfive.Downloader instance.")

File ~/anaconda/envs/eimason-sunpy/lib/python3.9/site-packages/sunpy/util/parfive_helpers.py:30, in Downloader.__init__(self, *args, **kwargs)
     27 if headers:
     28     kwargs['headers'] = headers
---> 30 super().__init__(*args, **kwargs)

TypeError: __init__() got an unexpected keyword argument 'headers'
```

**Comment by nabobalis** [comment]
I thought I already pinned it?

**Review by dstansby — APPROVED** [review]

**Comment by dstansby** [comment]
This is good, but shouldn't it be targeting the `3.1` branch? Currently targeting `3.0`, and I can't see `3.1` in the list of options...

**Comment by nabobalis** [comment]
This is the 3.1 branch.

**Comment by dstansby** [comment]
Eh, why's it called `3.0` then?

**Pull request merged**
